.github/workflows/ci_pr.yml

```diff
@@ -0,0 +1,140 @@
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.6
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.7
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.8
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-coverage --with-xunit --xunit-file=nosetests.__uuid__.xml
+              --with-timer --cover-erase --cover-inclusive --cover-branches --cover-package=azurelinuxagent"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+    
+    name: "Python ${{ matrix.python-version }} Unit Tests"
+    runs-on: ubuntu-18.04
+
+    env:
+      PYLINTOPTS: ${{ matrix.PYLINTOPTS }}
+      PYLINTFILES: ${{ matrix.PYLINTFILES }}
+      NOSEOPTS: ${{ matrix.NOSEOPTS }}
+      PYTHON_VERSION: ${{ matrix.python-version }}
+
+    steps:
+    
+    - name: Checkout WALinuxAgent repo
+      uses: actions/checkout@v2
+
+    - name: Setup Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    
+    - name: Install dependencies
+      run: |
+        sudo env "PATH=$PATH" python -m pip install --upgrade pip
+        sudo env "PATH=$PATH" pip install -r requirements.txt
+        sudo env "PATH=$PATH" pip install -r test-requirements.txt
+
+    - name: Run pylint
+      run: |
+        pylint $PYLINTOPTS --jobs=0 $PYLINTFILES
+
+    - name: Test with nosetests
+      if: always()
+      run: |
+        ./ci/nosetests.sh || EXIT_CODE=$?
+        exit $EXIT_CODE
+
+    - name: Run Coverage
+      env:
+        MIN_COVERAGE_GOAL: 60
+      if: matrix.python-version == 3.8
+      run: |
+        echo looking for coverage files :
+        ls -alh | grep -i coverage
+        sudo env "PATH=$PATH" coverage combine coverage.*.data
+        sudo env "PATH=$PATH" coverage xml
+        COVERAGE_PERCENT=$(coverage report | tail -n 1 | rev | cut -d ' ' -f 1 | cut -c 2- | rev)
+        EXIT_CODE=1
+        echo Codecoverage : $COVERAGE_PERCENT%
+        if [ "$COVERAGE_PERCENT" -eq "$COVERAGE_PERCENT" ] && [ "$COVERAGE_PERCENT" -gt "$MIN_COVERAGE_GOAL" ]; then 
+          echo Pass coverage check.
+          EXIT_CODE=0;
+        else 
+          echo Failed coverage check. Found $COVERAGE_PERCENT%, expected minimum of $MIN_COVERAGE_GOAL%;
+          EXIT_CODE=1;
+        fi
+        exit $EXIT_CODE
+
+    - name: Upload Coverage
+      if: ${{ github.event_name == 'pull_request' }} && matrix.python-version == 3.8
+      uses: codecov/codecov-action@v1
+      with:
+        file: ./coverage.xml
```

Comment:
It is really hard to see which test failed.
Can we add one of the commands to post the result output ?
```
    - name: Publish Unit Test Results
      uses: EnricoMi/publish-unit-test-result-action@v1.6
      if: matrix.python-version == 3.8 && always()
      with:
        github_token: ${{ secrets.GITHUB_TOKEN }}
        files: nosetests.*.xml
```

Reply:
>if: matrix.python-version == 3.8 && always()

Why only post for 3.8? Or will it post for all runs?

Additionally, publishing data should be a separate job (both UTs and CodeCov)

Reply:
> Additionally, publishing data should be a separate job (both UTs and CodeCov)

It would be much harder to publish data in a separate job because different jobs don't share a file system; is there a reason why we can't just have a separate step of the same job? That way, we could run scripts on the same VM, and access any files generated by prior steps (e.g., the `nosetest` run)

Reply:
Ohh nevermind then, just a suggestion mainly for clarity nothing else. You can keep it as is :)

Reply:
Yeah, a separate job would certainly be more readable! I don't think there's a good way of talking across jobs, though, short of, e.g., uploading to blob storage as an intermediary, which seems excessive.

Reply:
These kind of job can only run once, so we need to upload the logs separately, then download them back into that task
it is doable, but usually if UT fails, they should fail everywhere so this is fine for now.
If we see issues were UT fails in specific version of python, we can try to implement multiple reports.

Reply:
>If we see issues were UT fails in specific version of python, we can try to implement multiple reports.

Unfortunately that is exactly what we see lol (like most of the times tests pass in py2 and fail in py3 and vice versa). Might as well publish tests for all runs? If it becomes too verbose or dirty maybe we can remove it then?

Reply:
Its not too hard
@kevinclark19a  you can follow documentation here : https://github.com/EnricoMi/publish-unit-test-result-action#use-with-matrix-strategy
Basically after every test in the matrix, you call `actions/upload-artifact@v2` with file `nosetests.*.xml` and then you create a new job with task `actions/download-artifact@v2` and `EnricoMi/publish-unit-test-result-action@v1.6` to publish

Reply:
ping on this (@kevinclark19a to enable this and see how it looks in the UI page)

Reply:
Quick question: could we use this method for code cov as well? I think that would also benefit from being a separate job. Or, at the very least, a part of the new, "Publish Unit Test Results" job, instead of the "Run Unit Tests" job.

Reply:
For v1 of GHA, I don't think we should be over-complicating with coverage. Let's have it as it is today, only for py3.8 (and not required for merging) and get GHA working first.

Reply:
Also, we should add `fail_ci_if_error: false` here to not fail the check if codecov returns < 60%

Reply:
I can sync offline with anyone who wants more info on this, but it turns out that there is some unforeseen, additional complexity introduced via the methods mentioned above. For now, I'm going to stick to implementing only the behavior that we had prior, on Travis CI. However, I've created a branch for exploring these methods. Hopefully, this is something we can return to on a later revision of the pipeline!